kmapper/cover.py

```diff
@@ -94,3 +93,104 @@ def find_entries(self, data, cube, verbose=0):
+                 overlap_perc=None,
+                 # Bounds parameters
+                 limits=None):
+        """
+        limits: Numpy Array (n_dim,2)
+            (lower bound, upper bound) for every dimension
+            If a value is set to np.float('inf'), the bound will be assumed to be the min/max value of the dimension
+            Instead, if limits == None, this class works identically to the original Cover class.
+            i.e.
+                [[min_1, max_1],
+                 [min_2, max_2],
+                 [min_3, max_3]]
+        """
+        Cover.__init__(self, n_cubes, perc_overlap, nr_cubes, overlap_perc)
+        self.limits = limits
+
+        # Check limits can actually be handled and are set appropriately
+        NoneType = type(None)
+        assert isinstance(self.limits, (list, np.ndarray, NoneType)), 'limits should either be an array or None'
+        if isinstance(self.limits, (list, np.ndarray)):
+            self.limits = np.array(self.limits)
+            assert self.limits.shape[1] == 2, 'limits should be (n_dim,2) in shape'
+
+    def define_bins(self, data):
+        """Returns an iterable of all bins in the cover.
+
+        Example
+        -------
+            If there are 4 cubes per dimension and 3 dimensions return the bottom left (origin) coordinates of 64 hypercubes, as a sorted list of Numpy arrays
+
+        Warning: This function must assume that the first column of data are indices.
+        """
+
+        indexless_data = data[:, 1:]
+
+        # If self.limits is array-like
+        if isinstance(self.limits, np.ndarray):
+            dump_arr = np.zeros(self.limits.shape)  # dump_arr is used so we can change the values of self.limits from None to the min/max
+            dump_arr[:,0] = np.min(indexless_data, axis=0)
+            dump_arr[:,1] = np.max(indexless_data, axis=0)
+            dump_arr[self.limits != np.float('inf')] = 0
+            self.limits[self.limits == np.float('inf')] = 0
+            bounds_arr = self.limits + dump_arr
+            """ bounds_arr[i,j] = self.limits[i,j] if self.limits[i,j] == inf
+                bounds_arr[i,j] = max/min(indexless_data[i]) if self.limits == inf """
+            bounds = (bounds_arr[:,0], bounds_arr[:,1])
+
+            # Check new bounds are actually sensible - do they cover the range of values in the dataset?
+            if not ((np.min(indexless_data, axis=0) >= bounds_arr[:,0]).all() or \
+                    (np.max(indexless_data, axis=0) <= bounds_arr[:,1]).all()):
+                warnings.warn('The limits given do not cover the entire range of the lens functions\n' +\
+                              'Actual Minima: %s\tInput Minima: %s\n' % (np.min(indexless_data, axis=0), bounds_arr[:,0])+ \
+                              'Actual Maxima: %s\tInput Maxima: %s\n' % (np.max(indexless_data, axis=0), bounds_arr[:,1]))
+
+        else:  # It must be None, as we checked to see if it is array-like or None in __init__
```

Comment:
It looks like this will only apply bounds if bounds are given, otherwise just behave as the base class.  What about making this a little more explicit and not duplicating logic.  

What do you think about something like 
```
if isinstance(self.limits, np.ndarray):
    <existing logic>
else:
    super().find_bins(<params>)
```

This won't work exactly, but I think could take advantage of the inheritance a little better as there is a lot of repeated logic.



```